docs/SETUP_CI_CD_PIPELINE.md

```diff
@@ -64,6 +64,97 @@ others in the Forks group.
 * Click the *Connect* button to create the mirror of the GitHub repository on
 GitLab.
 
+### Understanding environments
+
+Environmments are the foundation of the pipeline.
+There are used in two contexts:
+* When getting and setting variables
+* When deploying the code 
+
+
+#### When deploying the code
+
+There are two types of environments: development and production.
+
+| Environment name | Type | Purpose |
+| --- | --- | --- |
+| dev | development | a developer's local development machine where they create applications |
+| CI | development | an environment created and hosted on GitLab to run automated tests continuously upon every commits (Continuous Integration) |
+| staging | production | an environmment hosted on AWS cloud for final acceptance of a version of the web site product that's like the real live in every aspect |
+| live | production | the real live web site product hosted on AWS cloud |
+
+The local environment is on the developer's machine, that's what the dev environment refers to. 
+The CI environment is implictely created by the CI part of GitLab, that's where the code is deployed for the execution of the automated tests.
+The CI is a gate-keeper for the production environments: deployment to staging and live can only happen if the tests pass in CI.
+
+#### Getting and settings variables on Gitlab
+
+##### Environment attribute
+
+A functionality of Gitlab is to store environment variables, so that we can use them in our deployed applications.
+Because there is multiple deployment environments and the variables often differ from one to the other, 
+Gitlab variables can be categorised and into different environment which are:
+* dev
+* staging
+* live
+* All (or \*)
+
+The environment is associated upon creation to each variable as one of its attributes.
+
+>**Note:** there is no connection between the environment of variables and the deployment environments listed in previous section. 
+>By convention the `staging` and `live` environments for variables are associated with the `staging` and `live` deployments respectively. 
+>(i.e: a staging variable is only to be used on staging deployment environment, 
+>and a live variable is to be used only on live deployment environment).
+>`All` class of variables are needed in applications regardless of their deployment environments, 
+>while the `dev` class of variables are equally used on a developer's local environments and on CI deployment environment.
+
+Furthermore, variables have a hierarchal organisation that map to groups and projects.
+So that, when there are variables is needed by all developers with the same value, such variable can be defined in a parent group, 
+which allow and sub-groups or sub-projects to access it without having to define it.
+
+##### Group and projects The diagram below show the hierarchy we have in place.
```

Comment:
`The diagram below show the hierarchy we have in place.` should be in a separate line below, not part of the sub-title.